### PR TITLE
Display JUnit Tags in Dev UI (Continuous Testing)

### PR DESCRIPTION
**Pull request by ueberfuhr**
I strongly use JUnits `@Tag` annotation to organize my tests, and I think this could be helpful to display them in the Dev UI. So I added this feature and a non-persistent checkbox to hide the tags, if needed.

See these screenshots:
Display Tags Feature (Dark Mode)
Display Tags Feature (Light Mode)

(`Q` is the shortcut for the `io.quarkus.test.junit.QuarkusTest` tag, so the user can see which test is a Quarkus test, and which one isn't.)

**Hint:**
The Vaadin Grid seems to recycle components, so the `<qui-badge>` too. However, this component is not implemented recyclable, which leads to inconsistent rendering. I have created a `<qui-recyclable-badge>`, that we can remove when the original badge is fixed. (see PR: https://github.com/qomponent/qui-badge/pull/4)

What do you think - is this a helpful feature?

### CODE REVIEW

**Comment by gsmet** [comment]
Sounds like an interesting addition. Can you filter on them?

/cc @phillip-kruger

**Review by phillip-kruger — CHANGES_REQUESTED** [review]
Thanks @ueberfuhr - This is great. Let's wait for https://github.com/quarkusio/quarkus/pull/42997 to be merged and then rebase this PR on top of that, so that you can use the badge that you fixed from qomponent.

**Comment by ueberfuhr** [comment]
> Sounds like an interesting addition. Can you filter on them?

Currently - no. But I could give it a try, if you confirm.

## Suggested Solution

I would not filter the results in the UI, but allow to restrict the test execution to a couple of selected tags, which would increase test performance.

The Quarkus Test Runtime allows to set application properties to include and exclude tags:

```properties
quarkus.test.include-tags=...
quarkus.test.exclude-tags=...
```

This leeds to an initial configuration of the `JUnit Test Runner` that we could derive during runtime.

Further affected classes (first analyzation):
 - `io.quarkus.devui.deployment.menu.ContinuousTestingProcessor`
 - `io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService`
 - `io.quarkus.dev.testing.ContinuousTestingSharedStateManager.State`

**Comment by gsmet** [comment]
Oh my question was just about being able to filter on a tag in the UI if you have a lot of tests (i.e. you click on a tag and only the tests with this tag are displayed and then you can drop the filter). But I'm not sure it makes sense?

**Comment by ueberfuhr** [comment]
IMHO: Such a client-side filter would be nice, but searching then should not only be limited to tag(s), but also to test classes and test names. Such a search form would be a new requirement.

Executing only tests with a given tag would filter the results in the UI too, but the filter then is server-side and more performant when repeating the tests. But I'm not sure if this is part of this PR, or if I should do this in another PR as a separate feature.

In summary, I would say:
 - this PR should only deal with the tag rendering in the DEV UI
 - another Issue/PR might allow to only execute tests with a given tag (or given tags)
 - another Issue/PR might introduce a filter form in Dev UI

Who can decide? 😉

**Comment by phillip-kruger** [comment]
I agree. Let's get this PR in, and then we can look at more features. The client side filter should be easy to do.

**Comment by ueberfuhr** [comment]
https://github.com/quarkusio/quarkus/pull/42997 is merged now, so I'll fix my code ASAP (tomorrow morning).

**Comment by phillip-kruger** [comment]
Let me know when you are ready for a review

**Review by phillip-kruger — APPROVED** [review]
Thanks !

**Comment by phillip-kruger** [comment]
@ueberfuhr can you squash your commits into one please. Thanks

**Comment by phillip-kruger** [comment]
If there are no tags, can we disable the option to display tags and have the default to not display them ?

**Comment by ueberfuhr** [comment]
> If there are no tags, can we disable the option to display tags and have the default to not display them ?

Yes, we can. 🙂 However, this needed a little refactoring of the `QwcContinuousTesting` class, because all the information about the results (incl. filtering by `obj.test === true`) are necessary when rendering the menu bar. Have tested this with 2 of my projects.

If I only have tests without any tag, it looks like this:

without tags

As soon as I introduce at least one tag:

with tag

**Comment by ueberfuhr** [comment]
> @ueberfuhr can you squash your commits into one please. Thanks

Sure, done.

**Review by phillip-kruger — APPROVED** [review]

**Comment by ueberfuhr** [comment]
I did a small change based on the findings from https://github.com/quarkusio/quarkus/issues/43067.

**Comment by gsmet** [comment]
@brunobat not sure what's going on but `OpenTelemetryInjectionsTest` has started to become extremely flaky.

**Comment by brunobat** [comment]
> @brunobat not sure what's going on but `OpenTelemetryInjectionsTest` has started to become extremely flaky.

@gsmet created this: https://github.com/quarkusio/quarkus/pull/43142

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java [hunk 1]
```diff
@@ -42,11 +42,7 @@
 import org.junit.jupiter.api.TestTemplate;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.platform.commons.annotation.Testable;
-import org.junit.platform.engine.FilterResult;
-import org.junit.platform.engine.TestDescriptor;
-import org.junit.platform.engine.TestExecutionResult;
-import org.junit.platform.engine.TestSource;
-import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.*;
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java:45]: The only issue is that we got back to star imports here. I tried to push to your branch but can't.

Could you adjust (and ideally configure your IDE to avoid star imports)?

**Review by gsmet — CHANGES_REQUESTED** [review]
Nice work!

There's just one last micro thing and then it's good to go.

**Comment by maxandersen** [comment]
Oh nice one. Tags is underrated so nice to see it in Devui. 

Should work nicely with https://quarkus.io/guides/getting-started-testing#running-specific-tests

Maybe a thing for the list of devui improvements for insights (or future one) @phillip-kruger

**Comment by phillip-kruger** [comment]
Yes very cool. I'll show this on insights

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java (hunk 1 above)
Comment (gastaldi) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java:45]: Here is the diff if you need help: 

```suggestion
import org.junit.platform.engine.FilterResult;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestExecutionResult;
import org.junit.platform.engine.TestSource;
import org.junit.platform.engine.UniqueId;
```

Comment (ueberfuhr) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java:45]: Yes, you're right. I disabled star imports right after the first commits. Thanks for the hints.

**Comment by ueberfuhr** [comment]
@maxandersen @gsmet - for filtering tests for execution or for rendering, I've created https://github.com/quarkusio/quarkus/issues/43208 for a discussion and a decision. 🙂

**Review by gsmet — APPROVED** [review]
Awesome contribution, thanks!
